Save filees won't load correctly

**Describe the bug**
When you load a save file, the text below the "Import Save" button says it has imported, but nothing changed.

**To Reproduce**
Steps to reproduce the behavior:
1. Open the app
2. Click "Import Save"
3. Import any .w2bzip
4. Nothing happens

**Expected behavior**
I expected the save file to load my save and all my elements/objects.

**Screenshots**
![image](https://github.com/user-attachments/assets/f71cfe33-cc7b-433f-bd84-d88959b11223)
Note how the save says it has imported

**Desktop (please complete the following information):**
 - OS: Windows 11, (tried on Windows 10 as well)
 - Version 0.3.3
 - Iris XE Integrated Graphics

I am also facing same issue
Windows 11

Below is my logs

[2025-02-12T13:42:54.537] [INFO] main - Running Wav2Bar v0.3.3-beta
[2025-02-12T13:42:54.538] [WARN] main - Can't write in app's root folder. Writing in app data folder provided by the OS.
[2025-02-12T13:42:54.538] [INFO] main - Working directory: C:\Wav2Bar
[2025-02-12T13:42:54.538] [INFO] main - creating main renderer...
[2025-02-12T13:42:54.570] [INFO] main - main renderer created.
[2025-02-12T13:42:54.827] [DEBUG] main_renderer - Loading modules done.
[2025-02-12T13:42:54.827] [INFO] main_renderer - loaded default save.
[2025-02-12T13:42:54.827] [DEBUG] main_renderer - is export window: false
[2025-02-12T13:42:54.828] [INFO] main_renderer - screen set to 1280x720
[2025-02-12T13:42:54.828] [INFO] main_renderer - animation stopped.
[2025-02-12T13:42:54.829] [INFO] main_renderer - FPS set to 60
[2025-02-12T13:42:54.829] [INFO] main_renderer - screen set to 1280x720
[2025-02-12T13:42:54.829] [INFO] main_renderer - screen set to 1280x720
[2025-02-12T13:42:54.829] [INFO] main_renderer - initializing settings...
[2025-02-12T13:42:54.829] [INFO] main_renderer - user settings found.
[2025-02-12T13:42:54.829] [DEBUG] main - reading C:\Wav2Bar/user/settings/user_settings.json...
[2025-02-12T13:42:54.830] [DEBUG] main - reading C:\Wav2Bar/user/settings/user_settings.json done.
[2025-02-12T13:42:54.831] [INFO] main_renderer - saving settings...
[2025-02-12T13:42:54.831] [DEBUG] main - writing json file to C:\Wav2Bar/user/settings/user_settings.json.
[2025-02-12T13:42:54.831] [INFO] main_renderer - loading settings...
[2025-02-12T13:42:54.831] [INFO] main_renderer - loaded settings!
[2025-02-12T13:42:54.831] [INFO] main_renderer - settings loaded:
[2025-02-12T13:42:54.831] [INFO] main_renderer - {"save_version":1,"software_version_used":"0.3.3-beta","ffmpeg":{"ffmpeg_path":"C:\\Wav2Bar\\ffmpeg-7.1-essentials_build\\bin\\ffmpeg.exe","ffprobe_path":"C:\\Wav2Bar\\ffmpeg-7.1-essentials_build\\bin\\ffprobe.exe"}}
[2025-02-12T13:42:54.831] [DEBUG] main - reading ./assets/help/help.json...
[2025-02-12T13:42:54.832] [DEBUG] main - wrote json file to C:\Wav2Bar/user/settings/user_settings.json.
[2025-02-12T13:42:54.832] [DEBUG] main - reading ./assets/help/help.json done.
[2025-02-12T13:42:54.832] [INFO] main_renderer - settings saved!
[2025-02-12T13:42:57.347] [DEBUG] main - reading directory C:\Users\susan
[2025-02-12T13:42:57.352] [DEBUG] main - reading directory C:\Users\susan done
[2025-02-12T13:42:59.399] [DEBUG] main - reading directory 
[2025-02-12T13:42:59.401] [ERROR] main_renderer -  do not exists: Error: Error invoking remote method 'read-dir': Error: read-dir: impossible to read this directory: Error: ENOENT: no such file or directory, scandir ''
[2025-02-12T13:42:59.702] [DEBUG] main - reading directory S
[2025-02-12T13:42:59.703] [ERROR] main_renderer - S do not exists: Error: Error invoking remote method 'read-dir': Error: read-dir: impossible to read this directory: Error: ENOENT: no such file or directory, scandir 'C:\Program Files (x86)\Wav2Bar\app-0.3.3-beta\resources\app\S'
[2025-02-12T13:43:00.038] [DEBUG] main - reading directory S:
[2025-02-12T13:43:00.041] [DEBUG] main - reading directory S: done
[2025-02-12T13:43:02.176] [DEBUG] main - reading directory S:\M&M Local
[2025-02-12T13:43:02.179] [DEBUG] main - reading directory S:\M&M Local done
[2025-02-12T13:43:03.828] [INFO] main_renderer - Backing up currently opened save...
[2025-02-12T13:43:03.828] [INFO] main_renderer - generating save file...
[2025-02-12T13:43:03.828] [DEBUG] main - writing json file to C:\Wav2Bar/temp/current_save/data.json.
[2025-02-12T13:43:03.828] [INFO] main - creating save file at C:\Wav2Bar/temp/before_new_save_open.w2bzip
[2025-02-12T13:43:03.829] [INFO] main_renderer - save file generated!
[2025-02-12T13:43:03.829] [INFO] main_renderer - Closing audio context if any...
[2025-02-12T13:43:03.829] [INFO] main_renderer - Audio context closed.
[2025-02-12T13:43:03.829] [INFO] main_renderer - Loading the save...
[2025-02-12T13:43:03.829] [INFO] main - loading save file at S:\M&M Local\demowav.w2bzip
[2025-02-12T13:43:03.831] [DEBUG] main - wrote json file to C:\Wav2Bar/temp/current_save/data.json.
[2025-02-12T13:43:03.831] [ERROR] main - failed at creating the save file: Error: ENOENT: no such file or directory, stat 'C:\Wav2Bar\temp\current_save\data.json'
[2025-02-12T13:43:03.845] [INFO] main_renderer - reading the save...
[2025-02-12T13:43:03.845] [DEBUG] main - reading C:\Wav2Bar/temp/current_save/data.json...
[2025-02-12T13:43:03.851] [DEBUG] main - reading C:\Wav2Bar/temp/current_save/data.json done.
[2025-02-12T13:43:03.851] [INFO] main_renderer - applying save...
[2025-02-12T13:43:03.851] [INFO] main_renderer - animation stopped.
[2025-02-12T13:43:03.851] [INFO] main_renderer - FPS set to 60
[2025-02-12T13:43:03.852] [INFO] main_renderer - screen set to 1280x720
[2025-02-12T13:43:03.852] [INFO] main_renderer - Save loaded!
[2025-02-12T13:43:14.899] [WARN] main - opening C:\Wav2Bar\logs.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@ https://keepachangelog.com/en/1.0.0/
 - Fixed: False information display for number of points count (#84)
 - Fixed: Save files won't load correctly (#73)
     - Wav2Bar now correctly uses the appdata directory instead of picking the disk root directory. To recover old settings on Windows, copy the content of `C:\Wav2Bar` over to `C:\Users\USERNAME\AppData\Roaming\Wav2Bar` (you likely only want the `user` directory there).
+- Some log messages were lost in specific scenarios where the window could not be identified.
 
 ## 0.3.3-beta (2023-02-04)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ https://keepachangelog.com/en/1.0.0/
 ### Added
 
 - Wav2Bar now warns if it fails to load a save file due to corrupted internal JSON data.
+- Wav2Bar now warns if it is not able to write app data (which can lead to save failure for instance).
 
 ### Fixed
 
@@ -14,7 +15,7 @@ https://keepachangelog.com/en/1.0.0/
     - An automatic clean up of render frames is done at export, and additional info is provided.
 - Fixed: False information display for number of points count (#84)
 - Fixed: Save files won't load correctly (#73)
-    - Wav2Bar now correctly uses the appdata directory instead of picking the disk root directory. To recover old settings on Windows, copy the content of `C:\Wav2Bar` over to `C:\Users\USERNAME\AppData\Roaming\Wav2Bar` (you likely only want the settings there).
+    - Wav2Bar now correctly uses the appdata directory instead of picking the disk root directory. To recover old settings on Windows, copy the content of `C:\Wav2Bar` over to `C:\Users\USERNAME\AppData\Roaming\Wav2Bar` (you likely only want the `user` directory there).
 
 ## 0.3.3-beta (2023-02-04)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,17 @@ https://keepachangelog.com/en/1.0.0/
 
 ## 0.3.4-beta (????-??-??)
 
-### Changed
+### Added
+
+- Wav2Bar now warns if it fails to load a save file due to corrupted internal JSON data.
+
+### Fixed
+
+- Fixed: Temporary folder is never cleaned up (#69)
+    - An automatic clean up of render frames is done at export, and additional info is provided.
+- Fixed: False information display for number of points count (#84)
+- Fixed: Save files won't load correctly (#73)
+    - Wav2Bar now correctly uses the appdata directory instead of picking the disk root directory. To recover old settings on Windows, copy the content of `C:\Wav2Bar` over to `C:\Users\USERNAME\AppData\Roaming\Wav2Bar` (you likely only want the settings there).
 
 ## 0.3.3-beta (2023-02-04)
 

--- a/js/index.js
+++ b/js/index.js
@@ -164,9 +164,13 @@ class SaveHandler {
         ipcRenderer.once("finished-caching-save", async () => {
             //read data cached in ./temp/current_save
             imports.utils.CustomLog("info","reading the save...");
-    
-            const JSON_data = await ipcRenderer.invoke("read-json-file",`${working_dir}/temp/current_save/data.json`);
-            this._save_data = JSON.parse(JSON.stringify(JSON_data)); //copy data
+
+            try {
+                const JSON_data = await ipcRenderer.invoke("read-json-file",`${working_dir}/temp/current_save/data.json`);
+                this._save_data = JSON.parse(JSON.stringify(JSON_data)); //copy data
+            } catch (error) {
+                MessageDialog("error", "The save file data could not be read. The save file may be corrupted. To check out, change the extension to '.zip' and unzip it (make a backup!), and look for a data.json file.");
+            }
     
             //check version
             if (this._save_data.save_version > this._CURRENT_SAVE_VERSION) {

--- a/js/index.js
+++ b/js/index.js
@@ -1462,6 +1462,12 @@ function LoadModules() {
  */
 function InitPage(export_mode) {
 
+    ipcRenderer.invoke("encountered-write-issue").then((issue) => {
+        if (issue) {
+            MessageDialog("error","An issue was encountered while trying to write app data. If you are running a system with restricted permissions, it is recommended to use the portable zip version instead. In all cases, it is discouraged to proceed to avoid save loss.");
+        }
+    });
+
     //SETUP PROJECT AND PREPARE SAVE
     project = new Project(export_mode);
     project.save_handler = new SaveHandler();

--- a/main.js
+++ b/main.js
@@ -49,6 +49,7 @@ process.chdir(__dirname);
 //folder to write temp data and user data
 let working_dir;
 let cant_write_to_root = false;
+let encountered_write_issue = false;
 
 
 
@@ -365,20 +366,25 @@ function Init() {
     let path_user_settings = path.resolve(working_dir, "./user/settings");
     let path_logs = path.resolve(working_dir, "./logs");
 
-    //create temp directory
-    if (!fs.existsSync(path_temp)) fs.mkdirSync(path_temp);
-    //clear existing cache if files remains from the last execution
-    fsExtra.emptyDirSync(path_temp);
-    //recreate the temp hierarchy
-    if (!fs.existsSync(path_temp_render)) fs.mkdirSync(path_temp_render);
-    if(!fs.existsSync(path_temp_current_save)) fs.mkdirSync(path_temp_current_save);
-
-    //create user directory
-    if(!fs.existsSync(path_user)) fs.mkdirSync(path_user);
-    if(!fs.existsSync(path_user_settings)) fs.mkdirSync(path_user_settings);
-
-    //create logs directory
-    if(!fs.existsSync(path_logs)) fs.mkdirSync(path_logs);
+    try {
+        //create temp directory
+        if (!fs.existsSync(path_temp)) fs.mkdirSync(path_temp);
+        //clear existing cache if files remains from the last execution
+        fsExtra.emptyDirSync(path_temp);
+        //recreate the temp hierarchy
+        if (!fs.existsSync(path_temp_render)) fs.mkdirSync(path_temp_render);
+        if(!fs.existsSync(path_temp_current_save)) fs.mkdirSync(path_temp_current_save);
+    
+        //create user directory
+        if(!fs.existsSync(path_user)) fs.mkdirSync(path_user);
+        if(!fs.existsSync(path_user_settings)) fs.mkdirSync(path_user_settings);
+    
+        //create logs directory
+        if(!fs.existsSync(path_logs)) fs.mkdirSync(path_logs);
+    } catch (e) {
+        encountered_write_issue = true;
+        main_log.error(`Couldn't create the necessary directories: ${e}`);
+    }
 
 
 
@@ -430,6 +436,10 @@ ipcMain.handle("argv", () => {
 ipcMain.handle("is-export-win", async (event) => {
     if (!export_win) return false;
     else return event.sender.id === export_win.webContents.id;
+});
+
+ipcMain.handle("encountered-write-issue", async () => {
+    return encountered_write_issue;
 });
 
 

--- a/main.js
+++ b/main.js
@@ -346,7 +346,7 @@ function PreInit() {
         })
         .catch(() => {
             cant_write_to_root = true;
-            working_dir = path.resolve(app.getPath("appData"), "/Wav2Bar");
+            working_dir = path.join(app.getPath("appData"), "/Wav2Bar");
             if (!fs.existsSync(working_dir)) fs.mkdirSync(working_dir);
 
             Init();

--- a/main.js
+++ b/main.js
@@ -32,7 +32,7 @@ const ft = require("fourier-transform/asm");
 require("./node_modules/log4js/lib/appenders/stdout");
 require("./node_modules/log4js/lib/appenders/console");
 const log4js = require("log4js");
-let main_log, main_renderer_log, export_log;
+let main_log, fallback_log, main_renderer_log, export_log;
 // eslint-disable-next-line no-unused-vars
 const colors = require("colors");
 
@@ -403,6 +403,7 @@ function Init() {
     main_log = log4js.getLogger("main");
     main_renderer_log = log4js.getLogger("main_renderer");
     export_log = log4js.getLogger("export");
+    fallback_log = log4js.getLogger("fallback_renderer_log");
 
     main_log.info(`Running Wav2Bar v${software_version}`);
     if (cant_write_to_root) main_log.warn("Can't write in app's root folder. Writing in app data folder provided by the OS.");
@@ -439,37 +440,41 @@ ipcMain.handle("is-export-win", async (event) => {
  * @param {String} log
  */
 ipcMain.handle("log", (event, type, log) => {
-    var logger;
-    switch (event.sender.id) {
-        case win.webContents.id: logger = main_renderer_log; break;
-        case export_win.webContents.id: logger = export_log; break;
-        default:
-            logger = main_log;
-            main_log.warn(`receiving logs from an unknown renderer with id ${event.sender.id}!`);
-            break;
+    let logger;
+    let using_fallback = false;
+    if (win && win.webContents && win.webContents.id === event.sender.id) {logger = main_renderer_log;}
+    else if (export_win && export_win.webContents && export_win.webContents.id === event.sender.id) {logger = export_log;}
+    else {
+        logger = fallback_log;
+        using_fallback = true;
+    }
+
+    let log_msg = log;
+    if (using_fallback) {
+        log_msg = `[RENDERER ${event.sender.id}] ${log}`;
     }
 
     switch (type) {
         case "trace":
-            logger.trace(log);
+            logger.trace(log_msg);
             break;
         case "debug":
-            logger.debug(log);
+            logger.debug(log_msg);
             break;
         case "info":
-            logger.info(log);
+            logger.info(log_msg);
             break;
         case "log":
-            logger.log(log);
+            logger.log(log_msg);
             break;
         case "warn":
-            logger.warn(log);
+            logger.warn(log_msg);
             break;
         case "error":
-            logger.error(log);
+            logger.error(log_msg);
             break;
         case "fatal":
-            logger.fatal(log);
+            logger.fatal(log_msg);
             break;
     }
 });

